genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/S3ArchivalServiceImpl.java

```diff
@@ -0,0 +1,172 @@
+     * target location - s3://$bucketName/run
+     * <p>
+     * following structure is created in S3
+     * s3://$bucketName/abc -> contents of run file
+     *
+     * @param file              file/directory to archive.
+     * @param targetLocationURI target location uri for the file
+     * @throws ArchivalException if archival fails
+     */
+
+    @Override
+    public void archive(
+        @NotNull final File file,
+        @NotNull final URI targetLocationURI
+    ) throws ArchivalException {
+        log.info("Archiving to location: {}", targetLocationURI);
+
+        try {
+            final AmazonS3URI targetLocationAmazonS3URI = new AmazonS3URI(targetLocationURI);
+            archiveRecursively(file, targetLocationAmazonS3URI);
+        } catch (Exception e) {
+            log.info("Error archiving to target location: {} ", targetLocationURI);
+            throw new ArchivalException("Error archiving file: " + file.getName(), e);
+        }
+    }
+
+    /**
+     * Traverse a directory recursively and archive the files and directories
+     * underneath it. If the directory is empty, create an empty object in S3.
+     * For a regular child file create an object in S3 and copy the file contents.
+     * For a empty child directory create an empty object in S3
+     *
+     * @param fileToArchive     file/dir to archive.
+     * @param targetLocationURI target location URI for storing the file in S3.
+     */
+    private void archiveRecursively(
+        @NotNull final File fileToArchive,
+        @NotNull final AmazonS3URI targetLocationURI
+    ) {
+        final File[] children = fileToArchive.listFiles();
+
+        //children == null if fileToArchive is not a directory.
+        //children is a empty array if the directory is empty
+        if (children == null || children.length == 0) {
+            archiveFile(fileToArchive, targetLocationURI);
+        } else { //Found a non empty directory, archive recursively
+            for (File childFile : children) {
+                archiveRecursively(
+                    childFile,
+                    new AmazonS3URI(targetLocationURI + File.separator + childFile.getName())
+                );
+            }
+        }
+    }
+
+    /**
+     * Archive a single file by creating an object in S3. For a file store its contents
+     * For a dir create a object with empty content.
+     *
+     * @param fileToArchive     file/dir to archive
+     * @param targetLocationURI base location for storing the file in S3
+     */
+    private void archiveFile(
+        @NotNull final File fileToArchive,
+        @NotNull final AmazonS3URI targetLocationURI
+    ) {
+        if (!fileToArchive.isDirectory()) {
+            uploadFile(targetLocationURI.getBucket(), targetLocationURI.getKey(), fileToArchive);
+        } else {
+            createEmptyFolder(targetLocationURI.getBucket(), targetLocationURI.getKey());
+        }
+    }
+
+    //Upload a file to the given bucket and key.
```

Comment:
Why is comment not in javadoc format? I know it's a private method and it doesn't really matter but for consistency it seems like the right thing to do

Reply:
Will do